pilot/pkg/config/kube/gateway/controller_test.go

```diff
@@ -0,0 +1,179 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gateway
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/config/aggregate/fakes"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/resource"
+	"k8s.io/client-go/kubernetes/fake"
+	svc "sigs.k8s.io/service-apis/api/v1alpha1"
+)
+
+var (
+	gatewayClassSpec = &svc.GatewayClassSpec{
+		Controller: ControllerName,
+	}
+	gatewaySpec = &svc.GatewaySpec{
+		Class: "gateway-class",
+	}
+	gateway = &networking.Gateway{
+		Servers: []*networking.Server{
+			{
+				Port: &networking.Port{
+					Number:   443,
+					Name:     "https",
+					Protocol: "HTTP",
+				},
+				Hosts: []string{"*.secure.example.com"},
+			},
+		},
+	}
+	httpRouteSpec = &svc.HTTPRouteSpec{}
+)
+
+func TestListInvalidGroupVersionKind(t *testing.T) {
+	g := NewGomegaWithT(t)
+	clientSet := fake.NewSimpleClientset()
+	store := &fakes.ConfigStoreCache{}
+	controller := NewController(clientSet, store)
+
+	typ := resource.GroupVersionKind{Kind: "wrong-kind"}
+	c, err := controller.List(typ, "ns1")
+	g.Expect(c).To(HaveLen(0))
+	g.Expect(err).To(HaveOccurred())
+}
+
+func TestList(t *testing.T) {
+	g := NewGomegaWithT(t)
+
+	clientSet := fake.NewSimpleClientset()
+	store := &fakes.ConfigStoreCache{}
+	controller := NewController(clientSet, store)
+
+	gwClassType := collections.K8SServiceApisV1Alpha1Gatewayclasses.Resource()
+	gwSpecType := collections.K8SServiceApisV1Alpha1Gateways.Resource()
+	gwType := collections.IstioNetworkingV1Alpha3Gateways.Resource()
+
+	store.ListReturnsOnCall(0, []model.Config{
+		{
+			ConfigMeta: model.ConfigMeta{
+				Type:      gwClassType.GroupVersionKind().Kind,
+				Group:     gwClassType.GroupVersionKind().Group,
+				Version:   gwClassType.GroupVersionKind().Version,
+				Name:      "gateway-class",
+				Namespace: "ns1",
+			},
+			Spec: gatewayClassSpec,
+		},
+	}, nil)
+	store.ListReturnsOnCall(1, []model.Config{
+		{
+			ConfigMeta: model.ConfigMeta{
+				Type:      gwSpecType.GroupVersionKind().Kind,
+				Group:     gwSpecType.GroupVersionKind().Group,
+				Version:   gwSpecType.GroupVersionKind().Version,
+				Name:      "gateway-spec",
+				Namespace: "ns1",
+			},
+			Spec: gatewaySpec,
+		},
+	}, nil)
+	c, err := controller.List(gwType.GroupVersionKind(), "ns1")
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(c).To(HaveLen(1))
```

Comment:
I don't think this really tests anything just checking the length is 1. Why not actually test the result is what we expect it to be?